File: 1283 Find the Smallest Divisor Given a Threshold/1283find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int ans = INT_MAX;
    
    void bin(int s,int e,int h,vector<int> arr){
        while(s<=e){
            int mid = s+(e-s)/2;
            long long int a1 = 0;
            for(int i=0;i<arr.size();i++){  
                a1+=((arr[i]+mid-1)/mid);
            }
            if(a1<=h){
                ans = min(ans,mid);
                e = mid-1;
            }
            else if(a1>h)
                s = mid+1;
        }
    }

    int smallestDivisor(vector<int>& nums, int threshold) {
        sort(nums.begin(),nums.end());
        int s = 1;
        int e = nums[nums.size()-1];
        bin(s,e,threshold,nums);
        return ans;
    }
};

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
import math
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        
        def cal(ele):
            ans = 0
            for num in nums:
                ans += math.ceil(num / ele)

            return ans

        l = 1
        r = max(nums)+1

        while l < r:
            mid = l + (r-l)//2
            if cal(mid) > threshold:
                l = mid + 1
            else:
                r = mid

        return l



